File: batch_translate.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
批量翻译 plugin-docs 目录下的 Markdown 文档
将 VSCode 术语替换为 Baosky，extension 替换为插件，并翻译为中文
"""

import os
import re
from pathlib import Path
import anthropic
import time

# API 密钥从环境变量获取
API_KEY = os.environ.get("ANTHROPIC_API_KEY", "")

# 专业术语映射（在翻译前应用）
TERM_REPLACEMENTS = {
    "VS Code": "Baosky",
    "VSCode": "Baosky",
    "Visual Studio Code": "Baosky",
    "vscode": "baosky",
}

def extract_frontmatter(content):
    """提取并返回 frontmatter 和剩余内容"""
    if content.startswith("---"):
        parts = content.split("---", 2)
        if len(parts) >= 3:
            return f"---{parts[1]}---", parts[2]
    return "", content

def extract_code_blocks(content):
    """提取代码块，用占位符替换"""
    code_blocks = []
    pattern = r'(```[\s\S]*?```)'

    def replacer(match):
        code_blocks.append(match.group(1))
        return f"__CODE_BLOCK_{len(code_blocks)-1}__"

    content = re.sub(pattern, replacer, content)
    return content, code_blocks

def restore_code_blocks(content, code_blocks):
    """恢复代码块"""
    for i, block in enumerate(code_blocks):
        content = content.replace(f"__CODE_BLOCK_{i}__", block)
    return content

def extract_inline_code(content):
    """提取行内代码，用占位符替换"""
    inline_codes = []
    pattern = r'(`[^`\n]+?`)'

    def replacer(match):
        inline_codes.append(match.group(1))
        return f"__INLINE_{len(inline_codes)-1}__"

    content = re.sub(pattern, replacer, content)
    return content, inline_codes

def restore_inline_code(content, inline_codes):
    """恢复行内代码"""
    for i, code in enumerate(inline_codes):
        content = content.replace(f"__INLINE_{i}__", code)
    return content

def replace_terms(content):
    """替换 VSCode 相关术语为 Baosky"""
    for old, new in TERM_REPLACEMENTS.items():
        content = content.replace(old, new)
    return content

def translate_content(content, client):
    """使用 Claude API 翻译内容"""
    if not content.strip():
        return content

    prompt = f"""请将以下Markdown文档翻译为中文。要求：

1. 将"extension"及其变体（Extension, extensions等）翻译为"插件"
2. 保持技术术语的准确性和一致性：
   - workspace -> 工作区
   - command -> 命令
   - API -> API（保持）
   - configuration -> 配置
   - debug -> 调试
   - workbench -> 工作台
   - theme -> 主题
   - marketplace -> 市场
   - activation event -> 激活事件
   - contribution point -> 贡献点
   - manifest -> 清单
   - entry point -> 入口点
   - webview -> Webview（保持）

3. 保留占位符（__CODE_BLOCK_X__、__INLINE_X__）不要翻译
4. 保持 Markdown 格式
5. 翻译要流畅自然，符合中文习惯
6. 标题也要翻译
7. 保留URL链接和文件路径不翻译

请直接输出翻译后的内容，不要添加任何解释：

{content}"""

    try:
        message = client.messages.create(
            model="claude-sonnet-4-5-20250929",
            max_tokens=16000,
            messages=[
                {"role": "user", "content": prompt}
            ]
        )
        return message.content[0].text
    except Exception as e:
        print(f"翻译出错: {e}")
        return content

def process_file(file_path, client):
    """处理单个文件"""
    print(f"\n处理文件: {file_path.relative_to('/root/my_website/my-website/docs/plugin-docs')}")

    try:
        # 读取文件
        with open(file_path, 'r', encoding='utf-8') as f:
            original_content = f.read()

        # 1. 提取 frontmatter
        frontmatter, content = extract_frontmatter(original_content)

        # 2. 替换 VSCode 术语为 Baosky
        content = replace_terms(content)

        # 3. 提取代码块
        content, code_blocks = extract_code_blocks(content)

        # 4. 提取行内代码
        content, inline_codes = extract_inline_code(content)

        # 5. 翻译内容（分段处理，避免超出token限制）
        # 按段落分割
        paragraphs = content.split('\n\n')
        translated_paragraphs = []

        for i, para in enumerate(paragraphs):
            if para.strip():
                if i % 10 == 0:
                    print(f"  翻译进度: {i}/{len(paragraphs)} 段落")
                translated = translate_content(para, client)
                translated_paragraphs.append(translated)
                time.sleep(0.5)  # 避免API限流
            else:
                translated_paragraphs.append(para)

        translated_content = '\n\n'.join(translated_paragraphs)

        # 6. 恢复行内代码
        translated_content = restore_inline_code(translated_content, inline_codes)

        # 7. 恢复代码块
        translated_content = restore_code_blocks(translated_content, code_blocks)

        # 8. 组合 frontmatter 和翻译后的内容
        final_content = frontmatter + translated_content if frontmatter else translated_content

        # 9. 写回文件
        with open(file_path, 'w', encoding='utf-8') as f:
            f.write(final_content)

        print(f"  ✓ 完成")
        return True

    except Exception as e:
        print(f"  ✗ 错误: {e}")
        return False

def main():
    """主函数"""
    if not API_KEY:
        print("错误: 未设置 ANTHROPIC_API_KEY 环境变量")
        return

    client = anthropic.Anthropic(api_key=API_KEY)

    base_dir = Path("/root/my_website/my-website/docs/plugin-docs")

    # 获取所有 .md 文件
    all_files = sorted(base_dir.rglob("*.md"))

    print(f"找到 {len(all_files)} 个 Markdown 文件需要翻译")
    print("="*60)

    success_count = 0
    fail_count = 0

    for file_path in all_files:
        if process_file(file_path, client):
            success_count += 1
        else:
            fail_count += 1

    print("\n" + "="*60)
    print(f"翻译完成!")
    print(f"成功: {success_count} 个文件")
    print(f"失败: {fail_count} 个文件")
    print(f"总计: {len(all_files)} 个文件")

if __name__ == "__main__":
    main()

File: batch-translate.js
#!/usr/bin/env node
/**
 * 批量翻译 Markdown 文档脚本
 * 快速翻译所有剩余的英文文档
 */

const fs = require('fs');
const path = require('path');

// 定义需要翻译的文件列表（按优先级）
const FILES_TO_TRANSLATE = [
  // 第一批：高优先级
  'references/commands.md',
  'references/contribution-points.md',
  'references/extension-manifest.md',
  'working-with-extensions/publishing-extension.md',

  // 第二批：extension-guides
  'extension-guides/color-theme.md',
  'extension-guides/command.md',
  'extension-guides/custom-data-extension.md',
  'extension-guides/custom-editors.md',
  'extension-guides/debugger-extension.md',
  'extension-guides/file-icon-theme.md',
  'extension-guides/markdown-extension.md',
  'extension-guides/notebook.md',
  'extension-guides/product-icon-theme.md',
  'extension-guides/scm-provider.md',
  'extension-guides/task-provider.md',
  'extension-guides/telemetry.md',
  'extension-guides/testing.md',
  'extension-guides/tree-view.md',
  'extension-guides/virtual-documents.md',
  'extension-guides/virtual-workspaces.md',
  'extension-guides/web-extensions.md',
  'extension-guides/webview.md',
  'extension-guides/workspace-trust.md',

  // 第三批：AI 相关
  'extension-guides/ai/ai-extensibility-overview.md',
  'extension-guides/ai/chat-tutorial.md',
  'extension-guides/ai/chat.md',
  'extension-guides/ai/language-model-chat-provider.md',
  'extension-guides/ai/language-model-tutorial.md',
  'extension-guides/ai/language-model.md',
  'extension-guides/ai/mcp.md',
  'extension-guides/ai/prompt-tsx.md',
  'extension-guides/ai/tools.md',

  // 第四批：其他目录
  'references/document-selector.md',
  'references/icons-in-labels.md',
  'references/theme-color.md',
  'references/when-clause-contexts.md',
  'ux-guidelines/activity-bar.md',
  'ux-guidelines/command-palette.md',
  'ux-guidelines/context-menus.md',
  'ux-guidelines/editor-actions.md',
  'ux-guidelines/notifications.md',
  'ux-guidelines/panel.md',
  'ux-guidelines/quick-picks.md',
  'ux-guidelines/settings.md',
  'ux-guidelines/sidebars.md',
  'ux-guidelines/status-bar.md',
  'ux-guidelines/views.md',
  'ux-guidelines/walkthroughs.md',
  'ux-guidelines/webviews.md',
  'working-with-extensions/bundling-extension.md',
  'working-with-extensions/continuous-integration.md',
  'working-with-extensions/testing-extension.md',
  'language-extensions/embedded-languages.md',
  'language-extensions/language-configuration-guide.md',
  'language-extensions/language-server-extension-guide.md',
  'language-extensions/programmatic-language-features.md',
  'language-extensions/semantic-highlight-guide.md',
  'language-extensions/snippet-guide.md',
  'language-extensions/syntax-highlight-guide.md',
  'advanced-topics/extension-host.md',
  'advanced-topics/python-extension-template.md',
  'advanced-topics/remote-extensions.md',
  'advanced-topics/tslint-eslint-migration.md',
  'advanced-topics/using-proposed-api.md',
];

const BASE_DIR = '/root/my_website/my-website/docs/plugin-docs';

// 简单的翻译映射（常见短语）
const TRANSLATIONS = {
  // 标题和常见短语
  'Overview': '概述',
  'Getting Started': '入门',
  'Get Started': '开始',
  'Prerequisites': '前提条件',
  'Installation': '安装',
  'Usage': '使用',
  'Examples': '示例',
  'API Reference': 'API 参考',
  'Configuration': '配置',
  'Troubleshooting': '故障排除',
  'FAQ': '常见问题',
  'Next Steps': '下一步',
  'Learn More': '了解更多',
  'See Also': '另请参阅',
  'Note:': '注意：',
  'Tip:': '提示：',
  'Warning:': '警告：',
  'Important:': '重要：',

  // 专业术语
  'extension': '插件',
  'Extension': '插件',
  'extensions': '插件',
  'Extensions': '插件',
  'workspace': '工作区',
  'Workspace': '工作区',
  'workspaces': '工作区',
  'Workspaces': '工作区',
  'debug': '调试',
  'Debug': '调试',
  'debugging': '调试',
  'Debugging': '调试',
  'command': '命令',
  'Command': '命令',
  'commands': '命令',
  'Commands': '命令',
  'activation event': '激活事件',
  'Activation Event': '激活事件',
  'activation events': '激活事件',
  'Activation Events': '激活事件',
  'contribution point': '贡献点',
  'Contribution Point': '贡献点',
  'contribution points': '贡献点',
  'Contribution Points': '贡献点',
  'manifest': '清单',
  'Manifest': '清单',
  'theme': '主题',
  'Theme': '主题',
  'themes': '主题',
  'Themes': '主题',
  'editor': '编辑器',
  'Editor': '编辑器',
  'terminal': '终端',
  'Terminal': '终端',
  'view': '视图',
  'View': '视图',
  'views': '视图',
  'Views': '视图',
  'panel': '面板',
  'Panel': '面板',
  'sidebar': '侧边栏',
  'Sidebar': '侧边栏',
  'status bar': '状态栏',
  'Status Bar': '状态栏',
  'activity bar': '活动栏',
  'Activity Bar': '活动栏',
  'marketplace': '市场',
  'Marketplace': '市场',
  'repository': '仓库',
  'Repository': '仓库',
  'package': '包',
  'Package': '包',
  'dependencies': '依赖项',
  'Dependencies': '依赖项',
  'settings': '设置',
  'Settings': '设置',
  'preferences': '首选项',
  'Preferences': '首选项',
};

/**
 * 检查文件是否包含大量英文内容
 */
function containsEnglish(content) {
  // 移除代码块
  const withoutCode = content.replace(/```[\s\S]*?```/g, '');
  // 移除内联代码
  const withoutInlineCode = withoutCode.replace(/`[^`]+`/g, '');
  // 移除 URL
  const withoutUrls = withoutInlineCode.replace(/https?:\/\/[^\s]+/g, '');

  // 检查是否有英文单词
  const englishWords = withoutUrls.match(/\b[a-zA-Z]{3,}\b/g);
  return englishWords && englishWords.length > 50;
}

/**
 * 简单翻译函数（仅用于标记已翻译的文件）
 */
function simpleTranslate(content) {
  // 这里不做实际翻译，只是添加中文标记
  // 实际翻译需要使用 AI 或人工翻译
  return content;
}

/**
 * 检查并报告文件状态
 */
function checkFile(filePath) {
  const fullPath = path.join(BASE_DIR, filePath);

  if (!fs.existsSync(fullPath)) {
    return { exists: false, needsTranslation: false };
  }

  const content = fs.readFileSync(fullPath, 'utf-8');
  const needsTranslation = containsEnglish(content);

  return {
    exists: true,
    needsTranslation,
    size: content.length,
    lines: content.split('\n').length
  };
}

/**
 * 主函数
 */
function main() {
  console.log('='.repeat(80));
  console.log('批量翻译文档检查');
  console.log('='.repeat(80));
  console.log();

  const stats = {
    total: FILES_TO_TRANSLATE.length,
    exists: 0,
    needsTranslation: 0,
    alreadyTranslated: 0,
    missing: 0
  };

  const needsTranslationList = [];

  FILES_TO_TRANSLATE.forEach((file, index) => {
    const result = checkFile(file);

    if (!result.exists) {
      stats.missing++;
      console.log(`[${index + 1}/${stats.total}] ❌ 文件不存在: ${file}`);
    } else {
      stats.exists++;
      if (result.needsTranslation) {
        stats.needsTranslation++;
        needsTranslationList.push(file);
        console.log(`[${index + 1}/${stats.total}] 📝 需要翻译: ${file} (${result.lines} 行)`);
      } else {
        stats.alreadyTranslated++;
        console.log(`[${index + 1}/${stats.total}] ✅ 已翻译: ${file}`);
      }
    }
  });

  console.log();
  console.log('='.repeat(80));
  console.log('统计摘要');
  console.log('='.repeat(80));
  console.log(`总文件数: ${stats.total}`);
  console.log(`存在的文件: ${stats.exists}`);
  console.log(`需要翻译: ${stats.needsTranslation}`);
  console.log(`已翻译: ${stats.alreadyTranslated}`);
  console.log(`缺失: ${stats.missing}`);
  console.log();

  if (needsTranslationList.length > 0) {
    console.log('需要翻译的文件列表:');
    needsTranslationList.forEach((file, index) => {
      console.log(`  ${index + 1}. ${file}`);
    });
  }
}

if (require.main === module) {
  main();
}

module.exports = { checkFile, containsEnglish };
